ravenframework/SupervisedLearning/SupervisedLearning.py

```diff
@@ -118,33 +147,116 @@ def __init__(self):
+    """
+      This function return the state of the ROM
+      @ In, None
+      @ Out, state, dict, it contains all the information needed by the ROM to be initialized
+    """
+    state = copy.copy(self.__dict__)
+
+    if state.get('featureSelectionAlgo') is not None:
+      del state['featureSelectionAlgo']
+    return state
+
+  def __setstate__(self, d):
+    """
+      Initialize the ROM with the data contained in newstate
+      @ In, d, dict, it contains all the information needed by the ROM to be initialized
+      @ Out, None
+    """
+    self.__dict__.update(d)
+    if self.saveParams:
+      fs = self.paramInput.findFirst("featureSelection")
+      if  fs is not None:
+        self.featureSelectionAlgo = featureSelectionFactory.returnInstance(fs.subparts[0].getName())
+        self.featureSelectionAlgo._handleInput(fs.subparts[0])
 
   def _handleInput(self, paramInput):
     """
       Function to handle the common parts of the model parameter input.
       @ In, paramInput, InputData.ParameterInput, the already parsed input.
       @ Out, None
     """
+    # check if paramInput must be saved.
+    # currently they are saved if and only if a feature selection
+    # algorithm is activated
+    # Consequentially we check for the presence of feature selection
+    # as first thing before processing the input data
+    featSelection = paramInput.findFirst("featureSelection")
+    self.saveParams = featSelection is not None
+    # now we can proceed with the input handling
     super()._handleInput(paramInput)
     nodes, notFound = paramInput.findNodesAndExtractValues(['Features', 'Target', 'pivotParameter'])
     assert(not notFound)
     self.features = nodes['Features']
     self.target = nodes['Target']
     self.pivotID = nodes['pivotParameter']
+
+    if featSelection is not None:
+      if len(featSelection.subparts) > 1:
+        self.raiseAnError(IOError, "Only one feature selection algorithm is allowed in the ROM")
+      featAlgo = featSelection.subparts[0]
+      self.featureSelectionAlgo = featureSelectionFactory.returnInstance(featAlgo.getName())
+      # handle input
+      self.featureSelectionAlgo._handleInput(featAlgo)
+      # if the feature selection algorithm is set, we should always have a mean to compute
+      # the feature importances (e.g. either the model can provide them or we use permutation)
+      self.computeImportances = True
+    # dim reduction?
+    dimReduction = paramInput.findFirst("dimensionalityReduction")
+    if dimReduction is not None:
+      self.performDimReduction = True
+      pca = dimReduction.findFirst("PCA")
+      if pca is None:
+        self.raiseAnError(IOError, "Only available dimensionality algorithms are: {}".format(",".join(['PCA'])))
+      nodesPCA, notFound = pca.findNodesAndExtractValues(['parametersToInclude', 'whichSpace', 'nComponents','whiten','tol'])
+      if nodesPCA['nComponents'] is not None: nodesPCA['nComponents'] = int(nodesPCA['nComponents'])
```

Comment:
Please put the "then" part on a separate line.
```
if nodesPCA['nComponents'] is not None: 
    nodesPCA['nComponents'] = int(nodesPCA['nComponents'])
```

Reply:
done